### PR TITLE
Introduced Selection#getSelectedBlocks()

### LINKED ISSUE #811: Try to extract "get selected blocks" logic from the list and heading features to the engine
Issue starter: https://github.com/ckeditor/ckeditor5-engine/issues/532#issuecomment-278900072

The code repeats in 2 places now and I need in the 3rd place (message quote), so it's time to hoist it.

### PR DESCRIPTION
**Pull request by Reinmar**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced `model.Selection#getSelectedBlocks()`. Closes #811.

---

### Additional information

Requires #815 (merged into this branch).

### CODE REVIEW

File: src/model/selection.js [hunk 1]
```diff
@@ -562,6 +562,50 @@ export default class Selection {
 	}
 
 	/**
+	 * Gets elements of type "block" touched by the selection.
+	 *
+	 * This method's result can be used for example to apply block styling to all
+	 * blocks covered by this selection.
+	 *
+	 * In this case the function will return all 3 paragraphs:
```
Comment (scofalik) [inline src/model/selection.js]: What exactly will be returned? Just 3 paragraphs? Or 3 paragraphs and quote? Clarification needed. If just 3 paragraphs it need to be clarified that quote is not seen as block.

File: src/model/selection.js [hunk 2]
```diff
@@ -661,3 +705,30 @@ export default class Selection {
 }
 
 mix( Selection, EmitterMixin );
+
+// Checks whether the given element extends $block in the schema and has a parent (is not a root).
+// Marks it as already visited.
+function checkIsBlockContainer( element, visited ) {
```
Comment (scofalik) [inline src/model/selection.js]: `function isUnvisitedBlockContainer( element, visited ) {`

File: src/model/selection.js [hunk 3]
```diff
@@ -661,3 +705,30 @@ export default class Selection {
...
+
+	// TODO https://github.com/ckeditor/ckeditor5-engine/issues/532#issuecomment-278900072.
+	// This should not be a `$block` check.
+	return element.document.schema.itemExtends( element.name, '$block' ) && element.parent;
+}
+
+// Finds the lowest element in position's ancestors which is a block.
+// Marks all ancestors as already visited to not include any of them later on.
+function getParentBlock( position, visited ) {
+	const ancestors = position.parent.getAncestors( { parentFirst: true, includeNode: true } );
```
Comment (scofalik) [inline src/model/selection.js]: `position.getAncestors()` should return same set of nodes

Comment (scofalik) [inline src/model/selection.js]: Sorry, I've been looking at the wrong version of code. Currently, `Position#getAncestors` returns them in parent-last order.

**Comment by scofalik** [comment]
What about such situation:
```
setData( doc, '<nestedBlock>a[<nestedBlock>b</nestedBlock>]</nestedBlock>' );
```
In this case, both `nestedBlock`s are returned. If we claim that nesting `$block`s is bad practice, then okay - as you've written in tests - this is edge case and it is not expected to work 100% correctly.

OTOH this might be fixable - we would have to check what are ranges start/end surroundings before deciding to push `range.start` or `range.end` parent to the results set.

**Comment by scofalik** [comment]
I've pushed commit fixing what I pointed out. If you don't want to do anything with the case I described above, feel free to merge this PR.

File: src/model/selection.js (hunk 2 above)
Comment (Reinmar) [inline src/model/selection.js]: Hm... To my taste, this is too verbose, but I understand your decision and can live with it :D.

**Comment by Reinmar** [comment]
> In this case, both nestedBlocks are returned. If we claim that nesting $blocks is bad practice, then okay - as you've written in tests - this is edge case and it is not expected to work 100% correctly.

Yeah, I knew that the cases I decided to cover were just the tip of the iceberg. There's a whole spectrum of weird schema configurations to cover and I decided to ignore them. It's never an easy choice because those things backfire sooner or later, so you need to trust yourself that the most common scenarios are covered.
